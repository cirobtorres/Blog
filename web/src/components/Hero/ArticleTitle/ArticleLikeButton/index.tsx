"use client";

import { useState } from "react";
import { useParams } from "next/navigation";
import { Popover } from "@/components/Shadcnui/popover";
import { PopoverTrigger } from "@radix-ui/react-popover";
import { dislikeArticle, likeArticle } from "../../../../service/articles";
import PopoverLoginContent from "@/components/Authentication/PopoverLoginContent";

function ArticleLikeButton({
  currentUser,
  totalLikes,
}: {
  currentUser: User;
  totalLikes: {
    documentId: string;
    users_permissions_user: {
      documentId: string;
    };
  }[];
}) {
  const { documentId: articleDocumentId }: { documentId: string } = useParams();
  const [likeCount, setLikeCount] = useState(totalLikes);
  const [articleLikeId, setArticleLikeId] = useState(
    totalLikes.filter(
      (like) =>
        like.users_permissions_user.documentId === currentUser.data?.documentId
    )[0]?.documentId
  );
  const [hasLiked, setHasLiked] = useState<boolean>(
    totalLikes.some(
      (like) =>
        like.users_permissions_user.documentId === currentUser.data?.documentId
    )
  );

  const formatLikeCount = (count: number) => {
    if (count <= 0) return "0 like";
    if (count === 1) return "1 like";
    return `${count} likes`;
  };

  if (currentUser.ok && currentUser.data) {
    const toggleLike = async () => {
      if (hasLiked) {
        setLikeCount((prevLikes) => {
          return prevLikes.filter(
            (like) =>
              like.users_permissions_user.documentId !==
              currentUser.data.documentId
          );
        });

        setHasLiked(!hasLiked);

        try {
          await dislikeArticle(articleLikeId);
        } catch (error) {
          console.error("Error while disliking article:", error);
          // Revert local state if update fails
          setLikeCount((prevLikes) =>
            hasLiked ? [...prevLikes] : prevLikes.slice(0, -1)
          );
          setHasLiked(hasLiked);
        }
      } else {
        setLikeCount((prevLikes) => {
          return [
            ...prevLikes,
            {
              documentId: currentUser.data.documentId,
              users_permissions_user: {
                documentId: currentUser.data.documentId,
              },
            },
          ];
        });

        setHasLiked(!hasLiked);

        try {
          const { data } = await likeArticle(
            articleDocumentId,
            currentUser.data.documentId
          );
          setArticleLikeId(data);
        } catch (error) {
          console.error("Error while liking article:", error);
          // Revert local state if update fails
          setLikeCount((prevLikes) =>
            hasLiked ? [...prevLikes] : prevLikes.slice(0, -1)
          );
          setHasLiked(hasLiked);
        }
      }
    };

    // Accessibility------------------------------------------------------------
    const handleKeyPress = (e: React.KeyboardEvent<HTMLButtonElement>) => {
      if (e.key === "Enter" || e.key === " ") {
        e.preventDefault();
        toggleLike();
      }
    };

    const buttonAriaLabel = hasLiked
      ? "Remover curtida do artigo"
      : "Curtir artigo";

    // currentUser.ok === true
    return (
      <div
        id="halb-user" // halb = hero-article-like-button
        data-testid="halb-user"
        className="flex items-center gap-4"
      >
        <button
          id="halb-button"
          data-testid="halb-button"
          type="button"
          onClick={toggleLike}
          onKeyDown={handleKeyPress}
          aria-pressed={hasLiked}
          aria-label={buttonAriaLabel}
          className="flex items-center gap-2"
        >
          <svg
            id="halb-icon-active"
            data-testid="halb-icon-active"
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
            fill="none"
            strokeWidth="2"
            strokeLinecap="round"
            strokeLinejoin="round"
            className={`lucide lucide-heart-icon lucide-heart size-6 text-transparent ${
              hasLiked
                ? "stroke-blog-foreground-highlight fill-blog-foreground-highlight"
                : "stroke-blog-foreground-readable fill-transparent"
            }`}
            aria-hidden="true"
          >
            <path d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z" />
          </svg>
          <p>{formatLikeCount(likeCount.length)}</p>
        </button>
      </div>
    );
  }

  // currentUser.ok === false
  return (
    <div id="halb" data-testid="halb" className="flex items-center gap-4">
      <Popover>
        <PopoverTrigger className="flex items-center gap-2">
          <svg
            id="halb-icon"
            data-testid="halb-icon"
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
            fill="none"
            strokeWidth="2"
            strokeLinecap="round"
            strokeLinejoin="round"
            className="lucide lucide-heart-icon lucide-heart size-6 stroke-blog-foreground-readable fill-transparent"
            aria-hidden="true"
          >
            <path d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z" />
          </svg>
          <p>{formatLikeCount(likeCount.length)}</p>
        </PopoverTrigger>
        <PopoverLoginContent align="center" />
      </Popover>
    </div>
  );
}

export default ArticleLikeButton;
