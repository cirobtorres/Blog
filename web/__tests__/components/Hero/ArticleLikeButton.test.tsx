import { faker } from "@faker-js/faker";
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import ArticleLikeButton from "../../../src/components/Hero/ArticleTitle/ArticleLikeButton";
import {
  createAuthenticatedUser,
  unauthorizedUserMock,
} from "../../../__mocks__/mockUser";
// import { randomInt } from "../../../__mocks__/utilities/randomInt";

faker.seed(1);

const length = 1;

const mockTotalLikes = () =>
  Array.from({ length }, () => ({
    documentId: faker.string.uuid(),
    users_permissions_user: {
      documentId: faker.string.uuid(),
    },
  }));

const authenticatedUserMock = createAuthenticatedUser();
const totalLikesMock = mockTotalLikes();

jest.mock("next/navigation", () => ({
  ...jest.requireActual("next/navigation"),
  useParams: jest
    .fn()
    .mockReturnValue({ documentId: "4136cd0b-d90b-4af7-b485-5d1ded8db252" }),
}));

jest.mock("../../../src/service/articles", () => ({
  likeArticle: jest.fn().mockResolvedValue({
    data: [
      {
        documentId: "4136cd0b-d90b-4af7-b485-5d1ded8db252",
        users_permissions_user: {
          documentId: "4136cd0b-d90b-4af7-b485-5d1ded8db252",
        },
      },
    ],
  }),
  dislikeArticle: jest.fn().mockResolvedValue({
    data: [
      {
        documentId: "4136cd0b-d90b-4af7-b485-5d1ded8db252",
      },
    ],
  }),
}));

jest.mock("../../../src/lib/graphQlClient", () => ({
  graphqlReadArticleClient: {
    request: jest.fn().mockResolvedValue({
      comments_connection: {
        pageInfo: {
          total: 3,
        },
      },
    }),
  },
}));

describe("ArticleLikeButton", () => {
  describe("User on", () => {
    it("renders in the document", async () => {
      render(
        <ArticleLikeButton
          currentUser={authenticatedUserMock}
          totalLikes={totalLikesMock}
        />
      );
      await act(async () => {
        const likeButtonContainer = screen.getByTestId("halb-user");
        const likeButton = screen.getByTestId("halb-button");
        const thumbsUpIcon = screen.getByTestId("halb-icon-active");
        expect(likeButtonContainer).toBeInTheDocument();
        expect(likeButton).toBeInTheDocument();
        expect(thumbsUpIcon).toBeInTheDocument();
        expect(thumbsUpIcon).toHaveAttribute("aria-hidden", "true");
      });
    });

    it("renders 0 like", async () => {
      render(
        <ArticleLikeButton
          currentUser={authenticatedUserMock}
          totalLikes={[]}
        />
      );
      await act(async () => {
        const likeButton = screen.getByTestId("halb-button");
        expect(likeButton).toHaveTextContent(`0 like`);
      });
    });

    it("renders 1 like", async () => {
      render(
        <ArticleLikeButton
          currentUser={authenticatedUserMock}
          totalLikes={[
            {
              documentId: faker.string.uuid(),
              users_permissions_user: {
                documentId: faker.string.uuid(),
              },
            },
          ]}
        />
      );
      await act(async () => {
        const likeButton = screen.getByTestId("halb-button");
        expect(likeButton).toHaveTextContent(`1 like`);
      });
    });

    it(`renders ${length} likes`, async () => {
      render(
        <ArticleLikeButton
          currentUser={authenticatedUserMock}
          totalLikes={totalLikesMock}
        />
      );
      await act(async () => {
        const likeButton = screen.getByTestId("halb-button");
        expect(likeButton).toHaveTextContent(`${length} like`);
      });
    });

    it("presses button like", async () => {
      render(
        <ArticleLikeButton
          currentUser={authenticatedUserMock}
          totalLikes={totalLikesMock}
        />
      );
      await act(async () => {
        const likeButton = screen.getByTestId("halb-button");
        const thumbsUpIcon = screen.getByTestId("halb-icon-active");
        // User haven't liked the article yet
        expect(likeButton).toHaveTextContent(`${length} like`);
        expect(likeButton).toHaveAttribute("aria-pressed", "false");
        expect(likeButton).toHaveAttribute("aria-label", "Curtir artigo");
        expect(thumbsUpIcon).toHaveClass("fill-blog-foreground-readable");
        // User likes the article
        fireEvent.click(likeButton);
        await waitFor(() =>
          expect(likeButton).toHaveTextContent(`${length + 1} likes`)
        );
        expect(likeButton).toHaveAttribute("aria-pressed", "true");
        expect(likeButton).toHaveAttribute(
          "aria-label",
          "Remover curtida do artigo"
        );
        expect(thumbsUpIcon).toHaveClass("fill-blog-foreground-highlight");
        // User dislikes the article
        fireEvent.click(likeButton);
        await waitFor(() =>
          expect(likeButton).toHaveTextContent(`${length} like`)
        );
        expect(likeButton).toHaveAttribute("aria-pressed", "false");
        expect(likeButton).toHaveAttribute("aria-label", "Curtir artigo");
        expect(thumbsUpIcon).toHaveClass("fill-blog-foreground-readable");
      });
    });

    it("matches the snapshot", async () => {
      const { asFragment } = render(
        <ArticleLikeButton
          currentUser={authenticatedUserMock}
          totalLikes={totalLikesMock}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("User off", () => {
    it("renders in the document", async () => {
      render(
        <ArticleLikeButton
          currentUser={unauthorizedUserMock}
          totalLikes={totalLikesMock}
        />
      );
      await act(async () => {
        const likeButtonContainer = screen.getByTestId("halb");
        expect(likeButtonContainer).toBeInTheDocument();
      });
    });

    it("matches the snapshot", async () => {
      const { asFragment } = render(
        <ArticleLikeButton
          currentUser={unauthorizedUserMock}
          totalLikes={totalLikesMock}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
