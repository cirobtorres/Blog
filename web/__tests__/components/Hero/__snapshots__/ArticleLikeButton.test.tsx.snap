// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleLikeButton User off matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-4"
    data-testid="halb"
    id="halb"
  >
    <button
      aria-controls="radix-:r1:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="flex items-center gap-2"
      data-state="closed"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-heart-icon lucide-heart size-6 stroke-blog-foreground-readable fill-transparent"
        data-testid="halb-icon"
        fill="none"
        height="24"
        id="halb-icon"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
        />
      </svg>
      <p>
        1 like
      </p>
    </button>
  </div>
</DocumentFragment>
`;

exports[`ArticleLikeButton User on matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-4"
    data-testid="halb-user"
    id="halb-user"
  >
    <button
      aria-label="Curtir artigo"
      aria-pressed="false"
      class="flex items-center gap-2"
      data-testid="halb-button"
      id="halb-button"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-heart-icon lucide-heart size-6 text-transparent stroke-blog-foreground-readable fill-transparent"
        data-testid="halb-icon-active"
        fill="none"
        height="24"
        id="halb-icon-active"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
        />
      </svg>
      <p>
        1 like
      </p>
    </button>
  </div>
</DocumentFragment>
`;
